src/aggregators/ccnrm.jl

```diff
@@ -0,0 +1,161 @@
+    CCNRMJumpAggregation{T, S, F1, affecttype, RNG, typeof(dg), typeof(ptt)}(
+        nj, nj, njt, et,
+        crs, sr, maj,
+        rs, affs!, sps,
+        rng, dg, ptt)
+end
+
++############################# Required Functions ##############################
+# creating the JumpAggregation structure (function wrapper-based constant jumps)
+function aggregate(aggregator::CCNRM, u, p, t, end_time, constant_jumps,
+        ma_jumps, save_positions, rng; kwargs...)
+
+    # handle constant jumps using function wrappers
+    rates, affects! = get_jump_info_fwrappers(u, p, t, constant_jumps)
+
+    build_jump_aggregation(CCNRMJumpAggregation, u, p, t, end_time, ma_jumps,
+        rates, affects!, save_positions, rng; num_specs = length(u),
+        kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::CCNRMJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    initialize_rates_and_times!(p, u, params, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+# execute one jump, changing the system state
+function execute_jumps!(p::CCNRMJumpAggregation, integrator, u, params, t, affects!)
+    # execute jump
+    u = update_state!(p, integrator, u, affects!)
+
+    # update current jump rates and times
+    update_dependent_rates!(p, u, params, t)
+    nothing
+end
+
+# calculate the next jump / jump time
+# just the first reaction in the first non-empty bin in the priority table
+function generate_jumps!(p::CCNRMJumpAggregation, integrator, u, params, t)
+    p.next_jump, p.next_jump_time = getfirst(p.ptt)
+
+    # Rebuild the table if no next jump is found. 
+    if p.next_jump == 0
+        binwidth = BINWIDTH_OVER_AVGTIME / sum(p.cur_rates)
+        min_time = minimum(p.ptt.times)
+        rebuild!(p.ptt, min_time, binwidth)
+        p.next_jump, p.next_jump_time = getfirst(p.ptt)
+    end
+
+    nothing
+end
+
+######################## SSA specific helper routines ########################
+
+# Recalculate jump rates for jumps that depend on the just executed jump (p.next_jump)
+function update_dependent_rates!(p::CCNRMJumpAggregation, u, params, t)
+    @inbounds dep_rxs = p.dep_gr[p.next_jump]
+    @unpack ptt, cur_rates, rates, ma_jumps, end_time = p
+    num_majumps = get_num_majumps(ma_jumps)
+
+    @inbounds for rx in dep_rxs
+        oldrate = cur_rates[rx]
+        times = ptt.times
+        oldtime = times[rx]
+
+        # update the jump rate
+        @inbounds cur_rates[rx] = calculate_jump_rate(ma_jumps, num_majumps, rates, u,
+            params, t, rx)
+
+        # Calculate new jump times for dependent jumps
+        if rx != p.next_jump && oldrate > zero(oldrate)
+            if cur_rates[rx] > zero(eltype(cur_rates))
+                update!(ptt, rx, oldtime, t + oldrate / cur_rates[rx] * (times[rx] - t))
+            else
+                update!(ptt, rx, oldtime, 2 * end_time)
```

Comment:
The problem is that in theory we should support a user changing `end_time` via a callback (or at least may want to at some point). So it would be better not to couple into it in that way.

Reply:
Maybe we could use floatmax then?

Reply:
Also, if the rate is zero though shouldn't it not get binned? Such reactions can't ever occur until they are updated to a non-zero rate.

Reply:
Right, anything that falls outside the current time window of the priority time table doesn't get binned, just gets passed over.

Reply:
`floatmax` is fine. Just make sure it has the same type as `t`.